### ISSUE TITLE
iOS Safari behavior

### ISSUE THREAD

**Issue by SamHall** [openseadragon/openseadragon#1618]
I'm having issues with iOS, specifically in Safari. 
Touching the canvas area causes the page to 'fire' a pseudo fullscreen mode, hiding the rest of the page content. 
On iPad, fullscreen cannot be interacted with consistently without unwanted 'jumping' behavior.
On iPhones, the top of the viewer is sometimes hidden under Safari's menu bar when firing fullscreen.

I'm trying to narrow down the issue but am hoping I've overlooked something here.

Specifically, I use Bootstrap (v3 & v4), and have not had issues until the past few months - with Safari on iOS. Any suggestions would be appreciated.
Thank you.

**Comment by iangilman** [comment]
Is this happening on your site now? Can you point to a page where it's happening? Does it happen for you on http://openseadragon.github.io/ ? 

I've just tested both the OSD site and your deepzoomchattanooga site, on both iPhone and iPad Safari with iOS 12.1.4. These are the things I've observed:

## iPad

* OSD and DZC seem to exhibit the same behaviors
* It's now possible to go into a true full screen! This wasn't supported in previous versions of iOS. That's kind of exciting! 
* The device's status bar is still visible at the top of the screen, so it overlaps the OSD buttons.
* iOS puts an X in the upper left to let you get out of full screen; this also overlaps the OSD buttons. 
* Some gestures, specifically dragging down from the top, pull you back out of full screen (which is annoying because you're probably just trying to pan). 
* After panning a few times I get an alert that says "It looks like you're typing while in full screen." Which of course I am not. 

## iPhone 

* It's still not possible to go to true full screen here, evidently; it's just doing the "full page" behavior.
* Everything looks fine on the OSD page, but the DZC page (at least https://chattanoogahistory.com/1889vs2018.php) sometimes has a little extra band of white above the viewer beneath the URL bar of the browser. Panning seems to get rid of this. 

Are these the things you're seeing or are you seeing additional/different stuff?

**Comment by SamHall** [comment]
Yes, the same 'hiding the buttons' behavior is present in several of the examples on github. (but not the first 'with default settings' or the custom toolbar - both of those work as expected).
Here's a test page I've set up to experiment with. If often takes the 2nd or third refresh/ attempt full-screen to see the buttons pushed under the ios/Safari menu bar.
[https://chattanoogahistory.com/safaritest](url)

**Comment by iangilman** [comment]
One thing I'm noticing on the iPhone is that the browser URL bar collapses itself when you scroll down the page, but you can expand it by tapping on it. On your test page I was unable to get the behavior you describe where the URL bar automatically overlaps the controls (perhaps because of your fix), but I could get it to happen manually: 

1. Scroll down the page so the URL bar collapses
2. Tap the full screen button; the controls are snug underneath the collapsed URL bar
3. Tap the URL bar; it expands so it's now overlapping the controls, which don't move down

So... seems like we have quite a number of issues here (including all the ones you've documented and the ones I have)! Probably the thing to do is to tackle them each individually... Which of them should be fixed inside OSD, and which ones are specific to your page? Are there facilities in iOS Safari to detect when these things are happening? What are the appropriate fixes?

Do you have any interest in working on fixing some of this stuff? If not, hopefully someone else will take an interest. Either way, it's good to have the issues documented!

**Comment by SamHall** [comment]
Yes, the expanding menu bar is the issue - good catch!
Ian, I definitely have the interest because it fuels my hobby/website, but seriously lack the skills other than helping test and comment. i.e. - I struggle with the slightest JavaScript. 
And I have you to thank for getting me as far as I have.

I will continue to experiment to see what's causing the issue in my more recent Bootstrap 4 pages, vs. the Bootstrap 3 pages.  I still think there's something 'not playing well' with Safari unique to those pages.
And - the issue is a bit elusive, not always showing itself. 
I'll help however I can within my abilities.  Thanks again!

**Comment by iangilman** [comment]
I'm guessing if it's due to the expanding menu bar, the intermittent nature has to do with some sort of timing thing, where when you go into fullscreen the menu bar is collapsed, but then a moment later it expands itself, but for some reason the web page doesn't realize it. You might look at things like resize events, or maybe Safari has some special events for things like the menu bar expanding or who knows. Another thing to look at would be to run a `setInterval` or something and measure the sizes of the window, the screen, various divs, etc. each time to see if you can detect a change.

There might also be CSS changes that could be made to make it more or less responsive. `position: fixed`, `height: 100vh`, things like that might be worth experimenting with.

My recommendation is to focus on tackling a single issue at a time. The expanding menu bar might be the one to start with. I'm happy to be a sounding board as you debug. :)

**Comment by Zapadore** [comment]
I am using Safari on my iPhone for an app I wrote (https://Enigmadevice.com/test2) and it works well except for two buttons at the top of the screen (where the navigation bar was) do not respond. I moved them down and now they work. Less than perfect.

**Comment by msalsbery** [comment]
This may be related to a Webkit bug, see #1962

Working on a workaround...
